File: dotnet/logbrew-dotnet/src/LogBrew/Traceparent.cs
using System;
using System.Collections.Generic;

namespace LogBrew
{
    public sealed class TraceparentContext
    {
        internal TraceparentContext(string version, string traceId, string parentSpanId, string traceFlags)
        {
            Version = version;
            TraceId = traceId;
            ParentSpanId = parentSpanId;
            TraceFlags = traceFlags;
            Sampled = (Convert.ToInt32(traceFlags, 16) & 1) == 1;
        }

        public string Version { get; }

        public string TraceId { get; }

        public string ParentSpanId { get; }

        public string TraceFlags { get; }

        public bool Sampled { get; }
    }

    public sealed class TraceparentSpanInput
    {
        private Dictionary<string, object?>? metadata;

        private TraceparentSpanInput(string name, string spanId, string status)
        {
            Name = name;
            SpanId = spanId;
            Status = status;
        }

        public string Name { get; }

        public string SpanId { get; }

        public string Status { get; }

        public double? DurationMs { get; private set; }

        public IReadOnlyDictionary<string, object?>? Metadata
        {
            get { return metadata; }
        }

        public static TraceparentSpanInput Create(string name, string spanId, string status)
        {
            return new TraceparentSpanInput(name, spanId, status);
        }

        public TraceparentSpanInput WithDurationMs(double durationMs)
        {
            DurationMs = durationMs;
            return this;
        }

        public TraceparentSpanInput WithMetadata(IDictionary<string, object?> value)
        {
            metadata = CopyMetadata(value);
            return this;
        }

        internal IDictionary<string, object?>? MetadataForSpan()
        {
            return metadata;
        }

        private static Dictionary<string, object?> CopyMetadata(IDictionary<string, object?> value)
        {
            var copied = Validation.CopyMetadata(value);
            var primitiveMetadata = new Dictionary<string, object?>();
            if (copied == null)
            {
                return primitiveMetadata;
            }

            foreach (var item in copied.Values)
            {
                primitiveMetadata[item.Key] = item.Value;
            }

            return primitiveMetadata;
        }
    }

    public static class Traceparent
    {
        private const string ZeroTraceId = "00000000000000000000000000000000";
        private const string ZeroSpanId = "0000000000000000";

        public static TraceparentContext Parse(string traceparent)
        {
            Validation.RequireNonEmpty("traceparent", traceparent);
            var normalized = traceparent.Trim().ToLowerInvariant();
            if (normalized.Length != 55 || normalized[2] != '-' || normalized[35] != '-' || normalized[52] != '-')
            {
                throw TraceparentShapeError();
            }

            var version = normalized.Substring(0, 2);
            var traceId = normalized.Substring(3, 32);
            var parentSpanId = normalized.Substring(36, 16);
            var traceFlags = normalized.Substring(53, 2);
            if (!IsLowerHex(version) || !IsLowerHex(traceId) || !IsLowerHex(parentSpanId) || !IsLowerHex(traceFlags))
            {
                throw TraceparentShapeError();
            }

            if (string.Equals(version, "ff", StringComparison.Ordinal))
            {
                throw new SdkException("validation_error", "traceparent version ff is forbidden");
            }

            RequireTraceId(traceId);
            RequireSpanId("traceparent parent span id", parentSpanId);
            RequireTraceFlags(traceFlags);
            return new TraceparentContext(version, traceId, parentSpanId, traceFlags);
        }

        public static string Create(string traceId, string spanId)
        {
            return Create(traceId, spanId, "01");
        }

        public static string Create(string traceId, string spanId, string? traceFlags)
        {
            var normalizedTraceId = NormalizeRequired("traceId", traceId);
            var normalizedSpanId = NormalizeRequired("spanId", spanId);
            var normalizedTraceFlags = NormalizeTraceFlags(traceFlags);
            RequireTraceId(normalizedTraceId);
            RequireSpanId("spanId", normalizedSpanId);
            RequireTraceFlags(normalizedTraceFlags);
            return "00-" + normalizedTraceId + "-" + normalizedSpanId + "-" + normalizedTraceFlags;
        }

        public static IReadOnlyDictionary<string, string> CreateHeaders(string traceId, string spanId)
        {
            return CreateHeaders(traceId, spanId, "01");
        }

        public static IReadOnlyDictionary<string, string> CreateHeaders(string traceId, string spanId, string? traceFlags)
        {
            return new Dictionary<string, string>
            {
                ["traceparent"] = Create(traceId, spanId, traceFlags)
            };
        }

        public static SpanAttributes SpanAttributesFromTraceparent(string traceparent, TraceparentSpanInput input)
        {
            if (input == null)
            {
                throw new SdkException("validation_error", "traceparent span input must be provided");
            }

            var context = Parse(traceparent);
            var spanId = NormalizeRequired("spanId", input.SpanId);
            RequireSpanId("spanId", spanId);
            var attributes = SpanAttributes.Create(input.Name, context.TraceId, spanId, input.Status)
                .WithParentSpanId(context.ParentSpanId);
            if (input.DurationMs.HasValue)
            {
                attributes.WithDurationMs(input.DurationMs.Value);
            }

            var metadata = input.MetadataForSpan();
            if (metadata != null)
            {
                attributes.WithMetadata(metadata);
            }

            return attributes;
        }

        private static string NormalizeRequired(string label, string value)
        {
            Validation.RequireNonEmpty(label, value);
            return value.Trim().ToLowerInvariant();
        }

        private static string NormalizeTraceFlags(string? traceFlags)
        {
            if (string.IsNullOrWhiteSpace(traceFlags))
            {
                return "01";
            }

            var providedTraceFlags = traceFlags ?? "01";
            return providedTraceFlags.Trim().ToLowerInvariant();
        }

        private static void RequireTraceId(string traceId)
        {
            if (traceId.Length != 32 || !IsLowerHex(traceId))
            {
                throw new SdkException("validation_error", "traceparent traceId must be 32 hex characters");
            }

            if (string.Equals(traceId, ZeroTraceId, StringComparison.Ordinal))
            {
                throw new SdkException("validation_error", "traceparent traceId must not be all zeros");
            }
        }

        private static void RequireSpanId(string label, string spanId)
        {
            if (spanId.Length != 16 || !IsLowerHex(spanId))
            {
                throw new SdkException("validation_error", label + " must be 16 hex characters");
            }

            if (string.Equals(spanId, ZeroSpanId, StringComparison.Ordinal))
            {
                throw new SdkException("validation_error", label + " must not be all zeros");
            }
        }

        private static void RequireTraceFlags(string traceFlags)
        {
            if (traceFlags.Length != 2 || !IsLowerHex(traceFlags))
            {
                throw new SdkException("validation_error", "traceparent traceFlags must be 2 hex characters");
            }
        }

        private static bool IsLowerHex(string value)
        {
            foreach (var character in value)
            {
                if (!((character >= '0' && character <= '9') || (character >= 'a' && character <= 'f')))
                {
                    return false;
                }
            }

            return true;
        }

        private static SdkException TraceparentShapeError()
        {
            return new SdkException(
                "validation_error",
                "traceparent must match W3C version-traceid-parentid-flags shape");
        }
    }
}

File: java/logbrew-java/src/main/java/co/logbrew/sdk/Traceparent.java
package co.logbrew.sdk;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Dependency-free helpers for explicit W3C traceparent interoperability.
 *
 * <p>The helpers validate and normalize incoming trace context, derive
 * LogBrew span attributes from app-owned child span IDs, and create outbound
 * traceparent carriers. They do not install OpenTelemetry or patch HTTP
 * clients.</p>
 */
public final class Traceparent {
    private static final Pattern TRACEPARENT_PATTERN = Pattern.compile(
        "^([0-9a-fA-F]{2})-([0-9a-fA-F]{32})-([0-9a-fA-F]{16})-([0-9a-fA-F]{2})$"
    );
    private static final Pattern TRACE_ID_PATTERN = Pattern.compile("^[0-9a-fA-F]{32}$");
    private static final Pattern SPAN_ID_PATTERN = Pattern.compile("^[0-9a-fA-F]{16}$");
    private static final Pattern FLAGS_PATTERN = Pattern.compile("^[0-9a-fA-F]{2}$");
    private static final String ZERO_TRACE_ID = "00000000000000000000000000000000";
    private static final String ZERO_SPAN_ID = "0000000000000000";

    private Traceparent() {
    }

    /**
     * Parsed W3C traceparent context after validation and normalization.
     */
    public static final class Context {
        private final String version;
        private final String traceId;
        private final String parentSpanId;
        private final String traceFlags;
        private final boolean sampled;

        private Context(String version, String traceId, String parentSpanId, String traceFlags) {
            this.version = version;
            this.traceId = traceId;
            this.parentSpanId = parentSpanId;
            this.traceFlags = traceFlags;
            this.sampled = (Integer.parseInt(traceFlags, 16) & 1) == 1;
        }

        /**
         * Returns the normalized W3C version field.
         */
        public String version() {
            return version;
        }

        /**
         * Returns the normalized 32-character trace ID.
         */
        public String traceId() {
            return traceId;
        }

        /**
         * Returns the normalized upstream parent span ID.
         */
        public String parentSpanId() {
            return parentSpanId;
        }

        /**
         * Returns the normalized two-character trace flags field.
         */
        public String traceFlags() {
            return traceFlags;
        }

        /**
         * Returns whether the W3C sampled bit is set.
         */
        public boolean sampled() {
            return sampled;
        }
    }

    /**
     * Inputs used to derive LogBrew span attributes from an incoming traceparent.
     */
    public static final class SpanInput {
        private final String name;
        private final String spanId;
        private final String status;
        private Double durationMs;
        private Map<String, ?> metadata;
        private List<SpanEventSummary> events;

        private SpanInput(String name, String spanId, String status) {
            this.name = name;
            this.spanId = spanId;
            this.status = status;
        }

        /**
         * Creates span input with required span identity and status fields.
         */
        public static SpanInput create(String name, String spanId, String status) {
            return new SpanInput(name, spanId, status);
        }

        /**
         * Sets optional span duration in milliseconds.
         */
        public SpanInput durationMs(double durationMs) {
            this.durationMs = Double.valueOf(durationMs);
            return this;
        }

        /**
         * Sets optional primitive metadata. Nested objects are rejected.
         */
        public SpanInput metadata(Map<String, ?> metadata) {
            this.metadata = Validation.copyMetadata(metadata);
            return this;
        }

        /**
         * Adds one optional privacy-bounded span event summary.
         */
        public SpanInput event(SpanEventSummary event) {
            if (event == null) {
                throw new SdkException("validation_error", "span event summary must be provided");
            }
            if (events == null) {
                events = new ArrayList<>();
            }
            events.add(event);
            SpanEventSummary.requireEventLimit(events.size());
            return this;
        }

        /**
         * Sets optional privacy-bounded span event summaries.
         */
        public SpanInput events(Iterable<SpanEventSummary> summaries) {
            if (summaries == null) {
                throw new SdkException("validation_error", "span events must be provided");
            }
            List<SpanEventSummary> copied = new ArrayList<>();
            for (SpanEventSummary summary : summaries) {
                if (summary == null) {
                    throw new SdkException("validation_error", "span event summary must be provided");
                }
                copied.add(summary);
                SpanEventSummary.requireEventLimit(copied.size());
            }
            this.events = copied;
            return this;
        }
    }

    /**
     * Parses, validates, and normalizes a W3C traceparent header value.
     */
    public static Context parse(String traceparent) {
        Validation.requireNonEmpty("traceparent", traceparent);
        Matcher matcher = TRACEPARENT_PATTERN.matcher(traceparent.trim());
        if (!matcher.matches()) {
            throw new SdkException(
                "validation_error",
                "traceparent must match W3C version-traceid-parentid-flags shape"
            );
        }

        String version = normalize(matcher.group(1));
        String traceId = normalize(matcher.group(2));
        String parentSpanId = normalize(matcher.group(3));
        String traceFlags = normalize(matcher.group(4));
        if ("ff".equals(version)) {
            throw new SdkException("validation_error", "traceparent version ff is forbidden");
        }
        requireTraceId(traceId);
        requireSpanId("traceparent parent span id", parentSpanId);
        requireTraceFlags(traceFlags);
        return new Context(version, traceId, parentSpanId, traceFlags);
    }

    /**
     * Creates a normalized sampled W3C traceparent header from explicit IDs.
     */
    public static String create(String traceId, String spanId) {
        return create(traceId, spanId, "01");
    }

    /**
     * Creates a normalized W3C traceparent header from explicit IDs and flags.
     */
    public static String create(String traceId, String spanId, String traceFlags) {
        String normalizedTraceId = normalizeRequired("traceId", traceId);
        String normalizedSpanId = normalizeRequired("spanId", spanId);
        String normalizedFlags = traceFlags == null || traceFlags.trim().isEmpty()
            ? "01"
            : normalize(traceFlags);
        requireTraceId(normalizedTraceId);
        requireSpanId("spanId", normalizedSpanId);
        requireTraceFlags(normalizedFlags);
        return "00-" + normalizedTraceId + "-" + normalizedSpanId + "-" + normalizedFlags;
    }

    /**
     * Creates a one-header outbound carrier containing only {@code traceparent}.
     */
    public static Map<String, String> createHeaders(String traceId, String spanId) {
        return createHeaders(traceId, spanId, "01");
    }

    /**
     * Creates a one-header outbound carrier containing only {@code traceparent}.
     */
    public static Map<String, String> createHeaders(String traceId, String spanId, String traceFlags) {
        return Collections.singletonMap("traceparent", create(traceId, spanId, traceFlags));
    }

    /**
     * Builds LogBrew span attributes that continue an incoming W3C traceparent.
     */
    public static SpanAttributes spanAttributesFromTraceparent(String traceparent, SpanInput input) {
        if (input == null) {
            throw new SdkException("validation_error", "traceparent span input must be provided");
        }
        Context context = parse(traceparent);
        String spanId = normalizeRequired("spanId", input.spanId);
        requireSpanId("spanId", spanId);
        SpanAttributes attributes = SpanAttributes
            .create(input.name, context.traceId(), spanId, input.status)
            .parentSpanId(context.parentSpanId());
        if (input.durationMs != null) {
            attributes.durationMs(input.durationMs.doubleValue());
        }
        if (input.metadata != null) {
            attributes.metadata(input.metadata);
        }
        if (input.events != null) {
            attributes.events(input.events);
        }
        return attributes;
    }

    private static String normalizeRequired(String label, String value) {
        Validation.requireNonEmpty(label, value);
        return normalize(value);
    }

    private static String normalize(String value) {
        return value.trim().toLowerCase(Locale.ROOT);
    }

    private static void requireTraceId(String traceId) {
        if (!TRACE_ID_PATTERN.matcher(traceId).matches()) {
            throw new SdkException("validation_error", "traceparent traceId must be 32 hex characters");
        }
        if (ZERO_TRACE_ID.equals(traceId)) {
            throw new SdkException("validation_error", "traceparent traceId must not be all zeros");
        }
    }

    private static void requireSpanId(String label, String spanId) {
        if (!SPAN_ID_PATTERN.matcher(spanId).matches()) {
            throw new SdkException("validation_error", label + " must be 16 hex characters");
        }
        if (ZERO_SPAN_ID.equals(spanId)) {
            throw new SdkException("validation_error", label + " must not be all zeros");
        }
    }

    private static void requireTraceFlags(String traceFlags) {
        if (!FLAGS_PATTERN.matcher(traceFlags).matches()) {
            throw new SdkException("validation_error", "traceparent flags must be two hex characters");
        }
    }
}
